Source/BSN.Commons/Infrastructure/MessageBroker/Kafka/KafkaEventAggregator.cs

```diff
@@ -0,0 +1,192 @@
+using System.Collections.Generic;
+using System.Threading;
+using Confluent.Kafka;
+using System.Threading.Tasks;
+using BSN.Commons.Infrastructure.MessageBroker.EventAggregator;
+using BSN.Commons.Infrastructure.MessageBroker.EventContracts.EventAggregator;
+using BSN.Commons.Infrastructure.MessageBroker.EventContracts.EventAggregator.Contracts;
+using BSN.Commons.Infrastructure.MessageBroker.EventContracts.EventAggregator.EventModels;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace BSN.Commons.Infrastructure.MessageBroker.Kafka
+{
+    /// <summary>
+    /// Represents an implementation of an event aggregator,
+    /// a fundamental component in an event-driven architecture,
+    /// responsible for aggregating events and broadcasting them to interested subscribers.
+    /// this implementation uses Kafka as the underlying message broker.
+    /// </summary>
+    public class KafkaEventAggregator : IEventAggregator
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="KafkaEventAggregator"/> class.
+        /// </summary>
+        /// <param name="connectionOptions">The Kafka connection options.</param>
+        /// <param name="subscriptionManager">The subscription manager that manages event subscriptions.</param>
+        /// <param name="logger">The logger.</param>
+        public KafkaEventAggregator(
+            KafkaConnectionOptions connectionOptions, 
+            IEventAggregatorSubscriptionManager subscriptionManager, 
+            ILogger logger
+            )
+        {
+            _logger = logger;
+            _producerConfig = new ProducerConfig
+            {
+                BootstrapServers = connectionOptions.BootstrapServers
+            };
+            _consumerConfig = new ConsumerConfig
+            {
+                BootstrapServers = connectionOptions.BootstrapServers,
+                GroupId = connectionOptions.ConsumerGroupId,
+                AutoOffsetReset = AutoOffsetReset.Earliest
+            };
+            _subscriptionManager = subscriptionManager ?? new InMemoryEventAggregatorSubscriptionManager();
+            _consumers = new Dictionary<string, IConsumer<Null, string>>();
+            _consumersCancellationTokenSources = new Dictionary<string, CancellationTokenSource>();
+        }
+
+        /// <inheritdoc />
+        public async void Publish<TEventModel>(IEvent<TEventModel> @event) where TEventModel : IEventDataModel
+        {
+            await PublishAsync(@event);
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync<TEventModel>(IEvent<TEventModel> @event) where TEventModel : IEventDataModel
+        {
+            // we did not do this in the constructor because we do not want to create a producer if we do not need it.
+            // right now for our use case, we only need to publish events or subscribe to them.
+            // we do not need both at the same time.
+            if (_producer == null)
```

Comment:
I suggest use producer factory instead of use directly

Reply:
doen